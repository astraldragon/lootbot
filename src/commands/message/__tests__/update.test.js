jest.mock("discord-akairo");
jest.mock("../../../models", () => {
  return {
    Message: { update: jest.fn("update") },
    Tier: { findOne: jest.fn("findOne") }
  };
});

const { Command } = require("discord-akairo");
const { Message, Tier } = require("../../../models");
const { checkManagePermissions } = require("../../../support");
const UpdateCommand = require("../update");

describe("commands/message/update", () => {
  beforeEach(() => {
    this.command = new UpdateCommand();

    this.args = {
      name: "msg1",
      message: null,
      type: null,
      delay: null,
      tier: null,
      user: null
    };

    this.tier = { id: 1, name: "Common" };
    this.user = { id: 1 };

    this.msg = {
      guild: { id: "1234" },
      channel: {
        send: jest.fn("send").mockReturnValue(Promise.resolve())
      }
    };
  });

  it("should configure properly", () => {
    expect(Command).toMatchSnapshot();
  });

  it("should validate the type field", () => {
    expect(Command.mock.calls[0][1].args[2].validate("nonvalid")).toBe(false);
    expect(Command.mock.calls[0][1].args[2].validate("intro")).toBe(true);
    expect(Command.mock.calls[0][1].args[2].validate("draw")).toBe(true);
    expect(Command.mock.calls[0][1].args[2].validate("reward")).toBe(true);
  });

  it("should have the proper permission check", () => {
    expect(Command.mock.calls[0][1].options.permissions).toBe(
      checkManagePermissions
    );
  });

  it("should update only non-null values", async () => {
    Message.update.mockReturnValue([{ id: 1 }, true]);
    this.args.delay = 10;
    await this.command.exec(this.msg, this.args);

    expect(Message.update).toHaveBeenCalledWith(
      { delay: 10 },
      { where: { name: this.args.name, guild: this.msg.guild.id } }
    );

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} updated`
    );
  });

  it("should query for the passed in tier", () => {
    this.args.tier = this.tier.name;
    this.command.exec(this.msg, this.args);

    expect(Tier.findOne).toHaveBeenCalledWith({
      where: { name: this.tier.name, guild: this.msg.guild.id }
    });
  });

  it("should notify when a tier isn't found", async () => {
    Tier.findOne.mockReturnValue(null);
    this.args.tier = this.tier.name;
    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      "A valid tier must be provided."
    );
  });

  it("should update the message with found tier", async () => {
    Tier.findOne.mockReturnValue(this.tier);
    Message.update.mockReturnValue([1]);

    this.args.tier = this.tier.name;
    await this.command.exec(this.msg, this.args);

    expect(Message.update).toHaveBeenCalledWith(
      { tier_id: this.tier.id },
      { where: { name: this.args.name, guild: this.msg.guild.id } }
    );

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} updated`
    );
  });

  it("should update a user if provided", async () => {
    Message.update.mockReturnValue([1]);

    this.args.user = this.user;
    await this.command.exec(this.msg, this.args);

    expect(Message.update).toHaveBeenCalledWith(
      { user_id: this.user.id },
      { where: { name: this.args.name, guild: this.msg.guild.id } }
    );

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} updated`
    );
  });

  it("should notify when the message does not exist", async () => {
    Message.update.mockReturnValue([0]);
    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} not found`
    );
  });

  it("should notify if any errors occur during the command", async () => {
    Message.update.mockImplementation(() => {
      throw new Error("test");
    });

    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `An error occurred updating ${this.args.name}`
    );
  });
});
