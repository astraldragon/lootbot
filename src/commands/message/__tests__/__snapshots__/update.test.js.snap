// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/message/update should configure properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "message-update",
      Object {
        "aliases": Array [
          "message-update",
          "mu",
        ],
        "args": Array [
          Object {
            "id": "name",
            "prompt": Object {
              "start": "What is the identier of the message?",
            },
            "type": "string",
          },
          Object {
            "default": null,
            "id": "message",
            "match": "prefix",
            "prefix": "message=",
            "prompt": Object {
              "optional": true,
              "start": "What is the message?",
            },
            "type": "string",
          },
          Object {
            "default": null,
            "id": "type",
            "match": "prefix",
            "prefix": "type=",
            "prompt": Object {
              "optional": true,
              "start": "What is the type of the message? (intro, draw, or reward)",
            },
            "type": "string",
            "validate": [Function],
          },
          Object {
            "default": null,
            "id": "delay",
            "match": "prefix",
            "prefix": "delay=",
            "prompt": Object {
              "optional": true,
              "start": "What is the delay until the next message is shown?",
            },
            "type": "integer",
          },
          Object {
            "default": null,
            "id": "tier",
            "match": "prefix",
            "prefix": "tier=",
            "prompt": Object {
              "optional": true,
              "start": "What is the tier of the loot? (blank will be included in all tiers of loot)",
            },
            "type": "string",
          },
          Object {
            "default": null,
            "id": "user",
            "match": "prefix",
            "prefix": "user=",
            "prompt": Object {
              "optional": true,
              "start": "What user should receive this message? (blank will be all users)",
            },
            "type": "member",
          },
        ],
        "category": "Message",
        "channelRestriction": "guild",
        "description": Object {
          "content": "Update message in the lootbox",
          "examples": Array [
            "message-update msg1 message=\\"New exciting message\\"",
            "message-update msg1 type=draw",
            "message-update msg1 delay=10 tier=Common",
            "message-update msg1 message=\\"You have won a <tier> prize...\\" user=@user",
          ],
          "usage": "<name> message=<message> type=<type> delay=<delay> tier=<tier> user=<user>",
        },
        "options": Object {
          "permissions": [Function],
        },
        "split": "sticky",
      },
    ],
  ],
}
`;
